runtime-rs: Implement resize memory for CH

cmaf opened the pull request:
Add resize_memory for Cloud Hypervisor and cloud_hypervisor_vm_resize for the call to the VM. Write unit tests checking failure cases.

Fixes #8801

pmores commented on src/runtime-rs/crates/hypervisor/src/ch/inner_hypervisor.rs:
```diff
@@ -761,17 +762,118 @@ impl CloudHypervisorInner {
+
+        if let Some(aligned_value_mb) = aligned_new_hotplug_size_mb {
+            let aligned_request = current_mem_size_mb + aligned_value_mb as u32;
+            if new_mem_mb != aligned_request {
```
Is this testing only done for the log message?  Because otherwise ~10 lines of code could be reduced to one.  (`new_mem_mb` will end up equal to `aligned_request` no matter what.)

pmores replied:
I don't mind logging this at all, I just wanted to make sure this is a conscious decision. Thanks!

pmores commented on src/runtime-rs/crates/hypervisor/src/ch/inner_hypervisor.rs:
```diff
@@ -761,17 +762,118 @@ impl CloudHypervisorInner {
+            if new_mem_mb != aligned_request {
+                info!(
+                    sl!(),
+                    "Aligning VM memory request {} to {}", new_mem_mb, aligned_request
```
Provided we have to have this log message, I'd propose including `block_size_mg` as well so that complete information related to aligning is available to anyone trying to track down a problem.

cmaf replied:
Fixed

pmores commented on src/runtime-rs/crates/hypervisor/src/ch/inner_hypervisor.rs:
```diff
@@ -761,17 +762,118 @@ impl CloudHypervisorInner {
+        if new_mem_mb == current_mem_size_mb {
+            info!(
+                sl!(),
+                "VM already has requested memory after alignment {}", new_mem_mb
```
It's a matter of taste but frankly, this wording seem a bit cryptic to me, considering that we're just trying to say that nothing needed to be done since the VM has had the requested amount of memory already.

cmaf replied:
I agree, it may not be necessary to log this.

cmaf commented on src/runtime-rs/crates/hypervisor/ch-config/src/ch_api.rs:
```diff
@@ -100,6 +100,31 @@ pub async fn cloud_hypervisor_vm_blockdev_add(
+        let response = simple_api_full_command_and_response(
+            &mut socket,
+            "PUT",
+            "vm.add-disk",
```
Yes. this was a copy/paste error, thank you for catching it!

cmaf commented:
@pmores thanks for the review. I updated a few things based on your comments and also modified some of the code comments for clarity. For logging, I chose to log the same information that is logged in the go runtime for cloud hypervisor. We could consider reducing it or switching it to debug.

lifupan commented on src/runtime-rs/crates/hypervisor/src/ch/inner_hypervisor.rs:
```diff
@@ -761,17 +762,116 @@ impl CloudHypervisorInner {
+            }
+        }
+
+        // Additional checks after memory alignment
```
Hi @cmaf 

Is the init default memory  block size aligned? If the initial values are aligned to the block size, then it is enough to do the alignment first and then do these size checks. Is there no need to check before the alignment?

pmores replied:
That's a very interesting question.  Correct me if I'm wrong but it would be hard to rely on `memory_info.default_memory` being aligned since 1) it comes from the config file, ie. the user, and 2) the memory allocation granularity (the block size) is only determined at runtime via the agent.

So I _think_ (and I might well be off here) that at least on the initial invocation of this function when `current_hotplug_size_mb` is likely zero and `current_mem_size_mb` thus consists purely of `memory_info.default_memory` alignment of the initial values can hardly be guaranteed.

That said, I still think you're right suggesting that the pre-alignment checks could be left out.  They seem to serve as an early-out to avoid the alignment computations if it's clear that they are not needed.  This would make a lot of sense if they were considerably resource-heavy.  However, since that doesn't seem the case I'd consider leaving the pre-alignment checks out, simplifying the function at the risk that a bunch of simple arithmetics might be performed unnecessarily at times.

cmaf replied:
@lifupan @pmores I switched the checks to only occur after alignment

pmores reviewed (approved):
lgtm, thanks @cmaf !

amshinde commented on src/runtime-rs/crates/hypervisor/src/ch/inner_hypervisor.rs:
```diff
@@ -761,17 +762,109 @@ impl CloudHypervisorInner {
+    ) -> Result<(u32, MemoryConfig)> {
+        // check for probe in memory_config
+        let caps = &self.capabilities;
+        if caps.is_mem_hotplug_probe_supported() {
```
Since we know that cloud-hypervisor supports memory hotplug, should be perform the check here?
I feel that this check needs to be done at a higher abstraction level (hypervisor level ) before this gets called at all.

amshinde replied:
Also since GuestMemoryProbe has not been added to the capability set yet, is this check passing currently? Or am I missing something?

pmores commented on src/runtime-rs/crates/hypervisor/src/ch/inner_hypervisor.rs:
```diff
@@ -761,17 +762,109 @@ impl CloudHypervisorInner {
+                );
+                new_mem_mb = aligned_request;
+            }
+        }
```
What happens if `aligned_new_hotplug_size_mb` is `None`, i.e. if aligning fails?  Do we want to continue with the old unaligned size?

pmores commented on src/runtime-rs/crates/hypervisor/src/ch/inner_hypervisor.rs:
```diff
@@ -761,17 +762,109 @@ impl CloudHypervisorInner {
 
     pub(crate) fn set_guest_memory_block_size(&mut self, size: u32) {
-        self._guest_memory_block_size_mb = size;
+        self.guest_memory_block_size_mb = size;
```
From my tests it looks like the value passed to this function is actually in bytes, not megabytes - you might want to double-check that...

pmores commented on src/runtime-rs/crates/hypervisor/src/ch/inner_hypervisor.rs:
```diff
@@ -761,17 +762,109 @@ impl CloudHypervisorInner {
 
-        Ok((0, MemoryConfig::default()))
+        // checked_next multiple_of() will align 0 to the block size, which we don't want here,
+        // so run check to avoid that
```
It seems that the function will do that not just for zero but for any input value that is aligned already.
